### DISCUSSION TITLE
Vweb https support

### DISCUSSION THREAD

**Discussion by khalyomede** [vlang/v#11866] [Ideas]
I am a vweb early enthusiast user and was wondering if it is possible for the binary file to serve over ssl? should I just generate an ssl certificate on my server?

**Comment by medvednikov** [comment]
You should use nginx for that :) It handles this very well.

This is not a priority for vweb at the moment, although contributions are welcome.

**Comment by khalyomede** [comment]
Thanks for your reply! In the mean time, if you allow me, I am pasting my Nginx configuration to do so. In case someone is passing by this issue and want to see an example of configuration (thanks @JalonSolov on Discord for leading me in the right direction):

_All the app resides in /home/alpine/dist folder._

```
.
├── dist/
│   ├── js/
│   │   └── index.js
│   └── index
└── src/
    └── index.v
```

 ```nginx
# /etc/nginx/conf.d/your-site.conf
server {
    listen 80;
    listen [::]:80;

    server_name localhost;

    error_log  /var/log/nginx/error.log;
    access_log /var/log/nginx/access.log;

    root /home/alpine/app/dist;

    gzip on;
    gzip_types text/html text/plain application/javascript application/x-javascript text/javascript text/xml text/css;

    # Authorize proxied responses to be compressed as well.
    gzip_proxied expired auth;

    # Zip only if the resource is above 1Kb, else the compressed file sometimes ends up bigger than the original file.
    gzip_min_length 1000;

    # Serve dynamic routes using V route handling
    location / {
        proxy_pass http://localhost:9000;
    }

    # Serve static files (JS and CSS, feel free to add more like PNG, ...).
    location ~ \.(js|css) {
        try_files $uri =404;
    }
}
```

```v
// src/index.v
import vweb

struct App {
  vweb.Context
}

fn (mut app App) index() vweb.Result {
  return app.html("<h1>Hello world</h1><script src='/js/index.js'></script>")
}

fn main() {
  vweb.run<App>(9000)
}
```

Finally, to add HTTPS, you can blindly follow these [Certbot instructions for Ubuntu Xenial Nginx](https://certbot.eff.org/lets-encrypt/ubuntuxenial-nginx.html) and you should be good to go (needs a domain name, so you will also want to edit `http://localhost:9000` by `https://my-domain.com:9000` and modify `server_name localhost` to `server_name my-domain.com`).

**Comment by JalonSolov** [comment]
You should run `v up`.  Your version of vweb is out of date... the run command is now `vweb.run<App>(&App{}, 9000)` as documented here:  https://modules.vlang.io/vweb.html

Thank you for the nginx config - that should help save some time for people attempting the same thing.

**Comment by medvednikov** [comment]
A simpler `vweb.run(&App{}, 9000)` also works, I've updated the docs.

**Comment by RolandHeinze** [comment]
I prefer to have TLS for TcpConn and UdpConn as well as HTTPS support in http. My use case is to write a local web server which runs as a windows service. One requirement is to use as little ressources (storage and RAM) as possible, and it should serve HTTPS because web browsers do not allow some kind of things without a secure connection. In that scenario it is not an option to use NGINX. Furthermore, the Go standard library offers a http package which supports HTTPS connections, and I think V should offer it, too.
Moreover, TLS for TcpConn is useful in much more scenarios (e.g. client wants to communicate with server over a secure channel).

**Comment by JalonSolov** [comment]
The `net` module does support https connections, and it handles them silently, based one whether the URL has `http` or `https` in front.

However, there is no support for _server_ side, because there is no implementation for handling server certificates, etc.  That's why nginx is recommended - it is reasonably small, and does have that support already.

It is possible V will gain that functionality, and if you want to work on it yourself, PRs are always welcome.  It just isn't a priority right now.
